ios/browser/api/bookmarks/importer/bookmarks_importer.cc

```diff
@@ -0,0 +1,178 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/api/bookmarks/importer/bookmarks_importer.h"
+
+#include "brave/ios/browser/api/bookmarks/importer/imported_bookmark_entry.h"
+#include "base/bind.h"
+#include "base/bind_helpers.h"
+#include "base/files/file_path.h"
+#include "base/base_paths.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/path_service.h"
+#include "components/bookmarks/browser/bookmark_model.h"
+#include "components/bookmarks/browser/bookmark_node.h"
+#include "components/bookmarks/browser/bookmark_utils.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "components/strings/grit/components_strings.h"
+#include "components/bookmarks/common/bookmark_pref_names.h"
+#include "ios/chrome/browser/application_context.h"
+#include "ios/chrome/browser/bookmarks/bookmark_model_factory.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state_manager.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "url/gurl.h"
+
+using bookmarks::BookmarkModel;
+using bookmarks::BookmarkNode;
+
+namespace {
+
+// Generates a unique folder name. If |folder_name| is not unique, then this
+// repeatedly tests for '|folder_name| + (i)' until a unique name is found.
+base::string16 GenerateUniqueFolderName(BookmarkModel* model,
+                                        const base::string16& folder_name) {
+  // Build a set containing the bookmark bar folder names.
+  std::set<base::string16> existing_folder_names;
+  const BookmarkNode* bookmark_bar = model->bookmark_bar_node();
+  for (const auto& node : bookmark_bar->children()) {
+    if (node->is_folder())
+      existing_folder_names.insert(node->GetTitle());
+  }
+
+  // If the given name is unique, use it.
+  if (existing_folder_names.find(folder_name) == existing_folder_names.end())
+    return folder_name;
+
+  // Otherwise iterate until we find a unique name.
+  for (size_t i = 1; i <= existing_folder_names.size(); ++i) {
+    base::string16 name = folder_name + base::ASCIIToUTF16(" (") +
+                          base::NumberToString16(i) + base::ASCIIToUTF16(")");
+    if (existing_folder_names.find(name) == existing_folder_names.end())
+      return name;
+  }
+
+  NOTREACHED();
+  return folder_name;
+}
+
+// Shows the bookmarks toolbar.
+void ShowBookmarkBar(ChromeBrowserState* browser_state) {
+  browser_state->GetPrefs()->SetBoolean(bookmarks::prefs::kShowBookmarkBar, true);
+}
+
+}  // namespace
+
+void BookmarksImporter::AddBookmarks(const base::string16& top_level_folder_name,
+                                     const std::vector<ImportedBookmarkEntry>& bookmarks) {
+  if (bookmarks.empty())
+    return;
+    
+  ios::ChromeBrowserStateManager* browser_state_manager =
```

Comment:
getting the browser state this way is a bad pattern to follow, the BookmarksImporter should be initialized with the BookmarkModel